src/building/dock.c

```diff
@@ -154,8 +247,48 @@ int building_dock_get_queue_destination(int ship_id, map_point *tile)
         }
         map_point_store_result(dock->x + dx, dock->y + dy, tile);
         if (!map_has_figure_at(map_grid_offset(tile->x, tile->y))) {
-            return dock_id;
+            if (!second_queue_available_dock) second_queue_available_dock = dock_id;
+            if (figure_trader_ship_can_queue_for_import(ship) && building_dock_can_import_from_ship(dock, ship_id)) {
+                importing_dock_id = dock_id;
+                // prioritize imports
+                break;
+            } else if (figure_trader_ship_can_queue_for_export(ship) && building_dock_can_export_to_ship(dock, ship_id)) {
+                exporting_dock_id = dock_id;
+                break;
+            }
         }
     }
+
+
+    if (importing_dock_id) {
+        return importing_dock_id;
+    } else if (exporting_dock_id) {
+        return exporting_dock_id;
+    }
+
+    if (ship->action_state == FIGURE_ACTION_110_TRADE_SHIP_CREATED) {
+        return first_queue_available_dock ? first_queue_available_dock : second_queue_available_dock;
+    }
     return 0;
 }
+
+int building_dock_get_can_trade_with_route(int route_id, int dock_id) {
+    building *dock = building_get(dock_id);
+    // for backword compatibility, avoiding default to reject all cities if game saved with older version
+    if (!dock->data.dock.has_accepted_route_ids) return 1;
+    return dock->data.dock.accepted_route_ids & (1 << route_id);
+}
+
+void building_dock_set_can_trade_with_route(int route_id, int dock_id, int can_trade) {
+    building *dock = building_get(dock_id);
+    if (!dock->data.dock.has_accepted_route_ids) {
+        dock->data.dock.has_accepted_route_ids = 1;
+        dock->data.dock.accepted_route_ids  = 0xffffffff;
+    }
+    int mask = 1 << route_id;
```

Comment:
@crudelios please double check this, I suspect it can break if we support more than 15 cities on x86 but not sure